StatsDownload/StatsDownload.TestHarness/TestHarnessStatsDownloadDatabaseProvider.cs

```diff
@@ -0,0 +1,108 @@
+﻿namespace StatsDownload.TestHarness
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Data.Common;
+    using System.Data.SqlTypes;
+    using System.Linq;
+    using Core.Interfaces;
+    using Core.Interfaces.DataTransfer;
+
+    public class TestHarnessStatsDownloadDatabaseProvider : IStatsDownloadDatabaseService
+    {
+        private readonly IStatsDownloadDatabaseService innerService;
+
+        private readonly ITestHarnessStatsDownloadSettings settings;
+
+        public TestHarnessStatsDownloadDatabaseProvider(IStatsDownloadDatabaseService innerService,
+            ITestHarnessStatsDownloadSettings settings)
+        {
+            this.innerService = innerService;
+            this.settings = settings;
+        }
+
+        public void Rollback(DbTransaction transaction)
+        {
+            innerService.Rollback(transaction);
+        }
+
+        public void Commit(DbTransaction transaction)
+        {
+            innerService.Commit(transaction);
+        }
+
+        public void FileDownloadError(FileDownloadResult fileDownloadResult)
+        {
+            innerService.FileDownloadError(fileDownloadResult);
+        }
+
+        public void FileDownloadFinished(FilePayload filePayload)
+        {
+            innerService.FileDownloadFinished(filePayload);
+        }
+
+        public DateTime GetLastFileDownloadDateTime()
+        {
+            return innerService.GetLastFileDownloadDateTime();
+        }
+
+        public bool IsAvailable()
+        {
+            return innerService.IsAvailable();
+        }
+
+        public void AddUsers(DbTransaction transaction, int downloadId, IEnumerable<UserData> usersData,
+            IEnumerable<FailedUserData> failedUsers)
+        {
+            innerService.AddUsers(transaction, downloadId, ModifiedUsers(usersData), failedUsers);
+        }
+
+        public IEnumerable<int> GetDownloadsReadyForUpload()
+        {
+            return innerService.GetDownloadsReadyForUpload();
+        }
+
+        public string GetFileData(int downloadId)
+        {
+            return innerService.GetFileData(downloadId);
+        }
+
+        public DbTransaction StartStatsUpload(int downloadId)
+        {
+            return innerService.StartStatsUpload(downloadId);
+        }
+
+        public void StatsUploadError(StatsUploadResult statsUploadResult)
+        {
+            innerService.StatsUploadError(statsUploadResult);
+        }
+
+        public void StatsUploadFinished(DbTransaction transaction, int downloadId, DateTime downloadDateTime)
+        {
+            innerService.StatsUploadFinished(transaction, downloadId, downloadDateTime);
+        }
+
+        public void NewFileDownloadStarted(FilePayload filePayload)
+        {
+            innerService.NewFileDownloadStarted(filePayload);
+        }
+
+        public void UpdateToLatest()
+        {
+            innerService.UpdateToLatest();
+        }
+
```

Comment:
Testing an exception occurring during the AddUsers method, will allow two inserts before exception thrown